packages/gatsby/src/utils/source-nodes.ts

```diff
@@ -74,23 +74,75 @@ function getStaleNodes(
       )
     }
 
+    if (state.statefulSourcePlugins.has(rootNode.internal.owner)) {
+      return false
+    }
+
     return !state.nodesTouched.has(rootNode.id)
   })
 }
 
 /**
- * Find all stale nodes and delete them
+ * Find all stale nodes and delete them unless the node type has been opted out of stale node garbage collection.
  */
-function deleteStaleNodes(
-  state: IGatsbyState,
-  nodes: GatsbyIterable<IGatsbyNode>
-): void {
-  const staleNodes = getStaleNodes(state, nodes)
+async function deleteStaleNodes(
+  previouslyExistingNodeTypeNames: Array<string>
+): Promise<void> {
+  const state = store.getState()
+
+  let deleteCount = 0
+
+  const cleanupStaleNodesActivity =
+    report.createProgress(`Clean up stale nodes`)
+
+  cleanupStaleNodesActivity.start()
+
+  const { pluginNamesToOwnedNodeTypes, statefulSourcePlugins } = state
+
+  const typeNamesToOwnerPluginName = new Map()
+
+  // build up a map of type names -> plugin names
+  pluginNamesToOwnedNodeTypes.forEach((ownedTypes, pluginName) => {
+    ownedTypes.forEach(typeName => {
+      typeNamesToOwnerPluginName.set(typeName, pluginName)
+    })
+  })
+
+  for (const typeName of previouslyExistingNodeTypeNames) {
+    const pluginName = typeNamesToOwnerPluginName.get(typeName)
 
-  staleNodes.forEach(node => store.dispatch(deleteNode(node)))
+    // no need to check this type if its owner has declared its a stateful source plugin
+    if (pluginName && statefulSourcePlugins.has(pluginName)) {
+      continue
+    }
+
+    report.verbose(`Checking for stale ${typeName} nodes`)
+
+    const nodes = getDataStore().iterateNodesByType(typeName)
+    const staleNodes = getStaleNodes(state, nodes)
+
+    for (const node of staleNodes) {
+      store.dispatch(deleteNode(node))
+      cleanupStaleNodesActivity.tick()
+
+      if (++deleteCount % 5000) {
+        // dont block event loop
+        await new Promise(res => {
+          setImmediate(() => {
+            res(null)
+          })
+        })
+      }
+    }
+  }
+
+  cleanupStaleNodesActivity.end()
+}
+
+export const is = {
+  initialSourceNodesOfCurrentNodeProcess: true,
```

Comment:
Only reason is to be able to mutate it during tests. Initially I added an extra env var to force checking stale nodes but I didn't like having a specific env var for it. 
I couldn't really think of another way to change this condition without mocking more of sourceNodes, but I didn't want to mock it since that's the code I'm testing. Any ideas on how I could make that better?

Reply:
Ah, just wrote another comment and then deleted it 😆 adding the comment makes sense, thank you!